Add Delete activity support for permanently deleted federated comments

obenland opened the pull request:
See https://github.com/Automattic/wordpress-activitypub/issues/2065.

## Summary

When a federated comment is permanently deleted using `wp_delete_comment()`, the plugin now creates a Delete activity and sends it to the appropriate ActivityPub inboxes. This ensures that remote servers are notified when comments they sent are deleted.

## Changes

- Added new hook `delete_comment` in `Comment` scheduler class
- Added `schedule_comment_delete_activity()` method that:
  - Checks if comment was previously federated using `\Activitypub\Comment::was_sent()`
  - Only creates Delete activities for federated comments
  - Ignores non-federated comments to avoid unnecessary federation traffic
- Added comprehensive tests with both positive and negative test cases
- Tests verify exact outbox post creation and metadata validation

## Testing instructions

1. Create a comment on a post by a registered user
2. Mark the comment as federated: `add_comment_meta( $comment_id, 'activitypub_status', 'federated' )`
3. Permanently delete the comment: `wp_delete_comment( $comment_id, true )`
4. Verify a Delete activity is created in the outbox with correct metadata
5. Test that non-federated comments do not create Delete activities

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>
<summary>Changelog Entry Details</summary>

#### Significance
- [x] Patch

#### Type
- [x] Added - for new features

#### Message
Add Delete activity support for permanently deleted federated comments.

</details>

obenland committed:
Add Delete activity support for permanently deleted federated comments

When a federated comment is permanently deleted using wp_delete_comment(),
the plugin now creates a Delete activity and sends it to the appropriate
ActivityPub inboxes. This ensures that remote servers are notified when
comments they sent are deleted.

Only comments that were previously federated (marked with 'federated'
meta status) will trigger Delete activities. Non-federated comments
are ignored to avoid unnecessary federation traffic.

matticbot committed:
Add changelog

obenland committed:
Fix test assertion logic for non-federated comment deletion

Compare total Delete activity counts before and after deletion,
not filtered results with unfiltered counts.

obenland commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -126,4 +126,118 @@ public function test_no_activity_scheduled( $comment_data ) {
+		);
+
+		$this->assertEmpty( $outbox_posts_after, 'Should not create Delete activity for non-federated comment deletion' );
+		$this->assertCount( count( $outbox_posts_before ), $outbox_posts_after, 'Number of Delete activities should remain unchanged' );
```
fixed in [e9f5378](https://github.com/Automattic/wordpress-activitypub/pull/2141/commits/e9f5378e7f51b57f9dbbb2fe87459e6a0603cb84).

obenland committed:
Simplify Delete activity tests with more focused queries

Reduce query complexity and remove unnecessary before/after comparisons.
Tests now specifically check for Delete activities with correct object IDs.

obenland commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -126,4 +126,133 @@ public function test_no_activity_scheduled( $comment_data ) {
+		\add_comment_meta( $comment_id, 'activitypub_status', 'federated' );
+
+		$comment        = \get_comment( $comment_id );
+		$activitypub_id = \Activitypub\Comment::generate_id( $comment );
```
Fixed in [4552d39](https://github.com/Automattic/wordpress-activitypub/pull/2141/commits/4552d39141af876df88958d7f64b815bdcca7060).

obenland commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -126,4 +126,133 @@ public function test_no_activity_scheduled( $comment_data ) {
+		);
+
+		$comment        = \get_comment( $comment_id );
+		$activitypub_id = \Activitypub\Comment::generate_id( $comment );
```
Fixed in [4552d39](https://github.com/Automattic/wordpress-activitypub/pull/2141/commits/4552d39141af876df88958d7f64b815bdcca7060).

obenland committed:
Merge branch 'trunk' into feature/send-delete-for-federated-comments

pfefferle commented:
Why not simply add `delete` to this check: https://github.com/Automattic/wordpress-activitypub/blob/073afc39df81daba761c2dc2a9cedc126e203448/includes/scheduler/class-comment.php#L62 ???

The hook is called anyways and it would make the code a bit simpler!?

obenland commented:
Sorry, I wasn't super explicit in the PR description with the context for this change, beyond mentioning #2065.

When a post gets deleted, [all associated comments also get deleted](https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/post.php#L3793-L3799), without going through the process of updating comment status.

That said, we could call `schedule_comment_activity()` internally, like in `schedule_comment_activity_on_insert()`.

pfefferle commented:
But `wp_delete_comment` is triggering [`wp_transition_comment_status`](https://developer.wordpress.org/reference/functions/wp_transition_comment_status/)!?

obenland commented:
Ha! Now your initial comment makes much more sense :) Yes, let's do that

pfefferle commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -7,6 +7,8 @@
 
 namespace Activitypub\Tests\Scheduler;
 
+use Activitypub\Comment;
```
```suggestion
use Activitypub\Collection\Outbox;
use Activitypub\Comment;
```

pfefferle commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -120,10 +122,106 @@ public function test_no_activity_scheduled( $comment_data ) {
+		// Check that no Delete activity was created for this specific comment.
+		$outbox_posts = \get_posts(
+			array(
+				'post_type'   => \Activitypub\Collection\Outbox::POST_TYPE,
```
```suggestion
				'post_type'   => Outbox::POST_TYPE,
```

pfefferle commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -120,10 +122,106 @@ public function test_no_activity_scheduled( $comment_data ) {
+		// Check if a Delete activity was created.
+		$outbox_posts = \get_posts(
+			array(
+				'post_type'   => \Activitypub\Collection\Outbox::POST_TYPE,
```
```suggestion
				'post_type'   => Outbox::POST_TYPE,
```

obenland committed:
Handle delete status through general comment activity

obenland committed:
Merge branch 'trunk' into feature/send-delete-for-federated-comments

pfefferle commented on tests/includes/scheduler/class-test-comment.php:
```diff
@@ -179,49 +179,4 @@ public function test_schedule_comment_delete_activity() {
 		$this->assertEquals( $activitypub_id, \get_post_meta( $outbox_post->ID, '_activitypub_object_id', true ) );
 		$this->assertEquals( self::$user_id, $outbox_post->post_author );
 	}
-
```
why not keep this test? it should have the same result!?

obenland replied:
I needed to update the test setup and add another check in the status transition handler. Fixed in [8c6f99a](https://github.com/Automattic/wordpress-activitypub/pull/2141/commits/8c6f99afd00b6a3a6d7463ecd3b505acce98be7b).

obenland committed:
Only sent deleted for replies to received comments

obenland committed:
Merge branch 'feature/send-delete-for-federated-comments' of https://github.com/Automattic/wordpress-activitypub into feature/send-delete-for-federated-comments

pfefferle commented on includes/scheduler/class-comment.php:
```diff
@@ -59,8 +59,8 @@ public static function schedule_comment_activity( $new_status, $old_status, $com
-			'trash' === $new_status ||
-			'spam' === $new_status
+			\in_array( $new_status, array( 'trash', 'delete', 'spam' ), true ) &&
+			\Activitypub\Comment::was_received( $comment->comment_parent )
```
We can not get sure that it is only a federated comment if the comment above was received. Comments directly to the post can also be federated for example or comments to federated local comments.

pfefferle replied:
Btw. there is already a check some lines below that is doing the trick: `should_comment_be_federated`.

obenland replied:
When comments get deleted, [all meta data is deleted](https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/comment.php#L1529-L1533) before `wp_transition_comment_status()` runs, so `Comment::was_sent()` won't know whether a comment was federated or not.

I guess we need to chose between sending Deletes out for all comments that get deleted or creating a separate hook callback that can check for it?

pfefferle replied:
It is in the same functions, only some lines below! https://github.com/Automattic/wordpress-activitypub/blob/4a1d3ea36c12bdd455f55a006662a44a6f92516c/includes/scheduler/class-comment.php#L73

pfefferle replied:
What I mean: It does not make any sense to have that check there, it has to be in the `should_comment_be_federated` function, otherwise it will break anyways.

pfefferle replied:
But you are right, the transition is set way too late, so maybe we should simply keep your function. Sorry :(

pfefferle replied:
What about reverting your changes, but instead of copying the code over, to simply call the `schedule_comment_activity` like `self::schedule_comment_activity( 'trash', 'publish', $comment )`???

obenland replied:
Maybe [45d4271](https://github.com/Automattic/wordpress-activitypub/pull/2141/commits/45d427160d187303fc544ba3c94f2b68e12c369d) is the charm?

obenland committed:
Reuse comment status transition hook

pfefferle commented on includes/scheduler/class-comment.php:
```diff
@@ -88,4 +90,17 @@ public static function schedule_comment_activity_on_insert( $comment_id, $commen
+	 */
+	public static function schedule_comment_delete_activity( $comment_id, $comment ) {
+		// Only send Delete activities for comments that were previously federated.
+		if ( \Activitypub\Comment::was_sent( $comment ) ) {
```
shoudln't we add this check to the `Comment::should_be_federated` instead?

obenland replied:
`was_sent` is  already part of `should_be_federated()` through `is_local()`. I decided to go with `was_sent()` here directly since that's what we're interested in and `should_be_federated()` runs as part of `schedule_comment_activity()`.

pfefferle commented on includes/scheduler/class-comment.php:
```diff
@@ -88,4 +90,17 @@ public static function schedule_comment_activity_on_insert( $comment_id, $commen
+	 */
+	public static function schedule_comment_delete_activity( $comment_id, $comment ) {
+		// Only send Delete activities for comments that were previously federated.
+		if ( \Activitypub\Comment::was_sent( $comment ) ) {
```
```suggestion
		if ( Comment_Utils::was_sent( $comment ) ) {
```

obenland replied:
[01351c1](https://github.com/Automattic/wordpress-activitypub/pull/2141/commits/01351c1ba27c9d69642948f38d0dc00c41c3da60)

obenland committed:
Use alias for Comment utils

Introduced an alias 'Comment_Utils' for the Activitypub\Comment class and updated usage in schedule_comment_delete_activity for improved code clarity.

Jiwoon-Kim commented:
https://core.trac.wordpress.org/ticket/63931#comment:10
